stock_picking_report_valued_sale_mrp/models/stock_move_line.py

```diff
@@ -0,0 +1,90 @@
+# Copyright 2020 Tecnativa - David Vidal
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+from odoo import api, fields, models
+
+
+class StockMoveLine(models.Model):
+    _inherit = 'stock.move.line'
+
+    phantom_product_id = fields.Many2one(
+        comodel_name="product.product",
+        compute="_compute_phantom_product_id",
+        compute_sudo=True,
+        string="Kit product",
+        readonly=True,
+    )
+    phantom_line = fields.Boolean(
+        compute="_compute_sale_order_line_fields",
+        compute_sudo=True,
+        default=False,
+    )
+    phantom_dlvd_qty = fields.Float(
+        compute="_compute_sale_order_line_fields",
+        compute_sudo=True,
+    )
+
+    @api.depends("sale_line")
+    def _compute_phantom_product_id(self):
+        """Relate every line with its kit product"""
+        for line in self.filtered(
+                lambda x: x.sale_line.product_id._is_phantom_bom()):
+            line.phantom_product_id = line.sale_line.product_id
+
+    def _compute_sale_order_line_fields(self):
+        """For kits we only want to store the value in one of the move lines to
+        avoid duplicate the amounts. We also need to recompute the total
+        amounts according to the corresponding delivered kits"""
+        super()._compute_sale_order_line_fields()
+        kit_lines = self.filtered("phantom_product_id")
+        for sale_line in kit_lines.mapped("sale_line"):
+            move_lines = kit_lines.filtered(
+                lambda x: x.sale_line == sale_line)
+            phantom_line = move_lines[:1]
+            price_unit = (
+                sale_line.price_subtotal / sale_line.product_uom_qty
+                if sale_line.product_uom_qty else sale_line.price_reduce)
+            # Get the kit. We should store the one used somewhere
+            bom = self.env['mrp.bom'].sudo()._bom_find(
```

Comment:
Isn't there any other mechanism? Why do we need to get the BoM? The idea is to get the information from the sale line. The problem if not is that there can be different BoMs when confirming the sales order than when printing the report.

Reply:
Yes, this is the main issue here. We don't have any info of the BoM used but we must figure out how many kits were delivered in the present picking.

If we assume the kits are inmutable (the components don't change in time), we could store them in the `sale.order.line` and then findout how many kits were delivered from the `stock.move.line`. Anyway, wouldn't that be a separate module?

Reply:
I don't think we need such bom linking. If you have the linked sale.order.line, we have the original kit quantity, and dividing by the whole pickings quantity (for considering backorders), we have the factor.

Reply:
Mmm, what happens if there are returns and redelivers?

Reply:
You can control that through the many2one field that points to the original/returned picking.

Reply:
I finding a big limitations with this aproach. Odoo can't tell very well how many kits were delivered when the picking is partial or a return is implied.

For example:

- I order 5 kits but I deliver just four and leave a backorder for the rest. Sale order line qty delivered = 0
- The I cancel the backorder: sale order line qty delivered = 5!

This limitation comes from the poor calculation that Odoo makes for these products:

https://github.com/odoo/odoo/blob/12.0/addons/sale_mrp/models/sale_mrp.py#L30-L33

Indeed, the aproach in v13 is precisely to link the boms with the moves: https://github.com/odoo/odoo/blob/13.0/addons/sale_mrp/models/sale_mrp.py#L25 and then compute according to the delivered component quantities.

Reply:
Changes done